### ISSUE TITLE
Enhancement suggestion: Compressed .ipynb files (e.g., .ipynbz)

### ISSUE THREAD

**Issue by callegar** [jupyterlab/jupyterlab#11824]
### Problem and proposed solution

The file format used by jupyter (`.ipynb`) can easily lead to very large files, particularly when images are embedded, and by its nature seems to be generally well compressible.

Rather than having authors compress `.ipynb` files with `gzip` or similar utilities or store them in `.zip` files for distribution, to then have users decompress them to view or use them, I wonder if it would be sensible to introduce an `.ipynbz` file format where compression is automatically and internally managed.

This would be in the lines of many other software projects that use text-based approaches (e.g., json, or xml) for their file formats but then introduce compression not to have too large files.

**Comment by davidbrochart** [comment]
Maybe this should be moved to https://github.com/jupyter/nbformat?

**Comment by callegar** [comment]
It is already there as https://github.com/jupyter/nbformat/issues/210, but along the last year I got the impression that everything is very still there, as the "format" is not seen as a moving target.

So I tried posting here to, as it is unclear to me if this really has to do with the format or just with some "envelope" for the format. 

In fact, if an `.ipynbz` format is introduced that is just a zstd-compressed `.ipynb` format, in principle this change can be applied just at the client level and selectively only on the more modern clients as a start. For clients not supporting compression, nothing changes wrt what you need to do today if you gzip or zstd-compress `.ipynb` files (that is, it is enough to decompress them to use them and to recompress them for distribution or achival). Modern clients would benefit from this becoming *transparent* to the user, though.

**Comment by callegar** [comment]
An alternative, possibly advantageous, but less compatible approach could be to move to a zip based format, where one has a zip envelope with the current ipynb content in it as a file, but with the embedded images moved to their own files inside the zip.  This would make it quite easy to support other types of attachments and to distribute sample data together with the notebooks. See https://github.com/jupyterlab/jupyterlab/issues/11825.

**Comment by bollwyvl** [comment]
This has been discussed on a number of occasions. I think the use case needs to be very carefully described.

As a _source_ format, the biggest, broadest complaint about the notebook format is complexity with bog-standard text tools like `git`(hub). Despite the fact that very lovely diffs _can_ be generated from a well-structured JSON format, it _isn't_ in the tools people use all the time. Any binary format would exacerbate this problem, taking interactive merge conflicts from being "very hard" to "never mind".

As an _archival_ format, rather than implementing a new one, PDF is very attractive. A single PDF can contain an arbitrary file attachment tree, as demonstrated on e.g. [notebook-as-pdf](https://github.com/betatim/notebook-as-pdf). Aside from being extremely portable, viewable, and predictable, PDF has some well-thought out advantages for archival and compliance, such as PDF A/2. But would be more expensive, computationally, than a straight zip.

As to the underlying breaking out e.g. sources and outputs as well-named individual files, this has been implemented by [nbexplode](https://github.com/takluyver/nbexplode), as well as to fit on distributed database e.g. [jupyterlab-dat](https://github.com/deathbeds/jupyterlab-dat).

As a _wire_ format: I think eventually we'll see `api/contents/*.ipynb` eventually get served by a much more granular api, probably implemented over piecemeal JSON websockets and plain-old-https-binary which is actually driven by what's on the screen. This is part of the ongoing work to support the CRDT-based collaboration.

**Comment by callegar** [comment]
@bollwyvl I agree with many of your points and this is why I am not suggesting abandoning the `.ipynb` format in favor of a compressed one, rather providing support for transparently opening and saving `.ipynb` files that have been passed through `gzip` (or, better, zstd). These `.ipynb.gz` or `.ipynb.zst` files could, for the sake of brevity, be called `.ipynbz`, but this is actually secondary.

Let me try to offer a better clarification:

> As a source format, the biggest, broadest complaint about the notebook format is complexity with bog-standard text tools like git(hub)
Understandable, but big triggers for this were probably actual bugs. I remember the one about all cell ids changing at every new save, making diffs sort of impossible.  Now, my proposal is not disrupting in this sense allowing one to either use the compressed or uncompressed .ipynb format depending on his needs. Furthermore, because the proposal is to add a standard compression supported by regular tools to normal `.ipynb` files, tools like `git diff` can be readily taught how to diff these files for human consumption (via the `textconv` attribute or by using an ad hoc external diff program). Even ad-hoc merge tools are possible.

> As an archival format, rather than implementing a new one, PDF is very attractive.

I think that a very good comparison for my enhancement proposal can be `.svg`. Here the format is a text based one (xml) not that dissimilar from the json used in notebooks. While the format is text-based, most tools support transparent compression and decompression of `.svg.gz` files, that are generally indicated as `.svgz` files. Should you want/need to process the XML or store it in a git archive, you always have the option to gunzip the `.svgz` files into `svg` ones.

Apart from that, yes, I agree that introducing a completely alternative notebook format based on a container format *with an index* would have many advantages, not last the possibility of having generalized attachments (e.g. attach a data set, an audio file, a video stream, meant to be processed by the notebook to the notebook itself). Zip based formats are now used by many tools with success (think Opendocument, docx, etc.).

**Comment by platise** [comment]
Using zip could, besides much shorter files, provide a beneficial feature to encrypt ipynb's at the same time with a password.

We run jupyter on a server and this feature would simplify access control a lot for a patent related research, instead of providing different password protected servers.
